Copy all `AsciiExt` methods to the primitive types directly in order to deprecate it later

LukasKalbertodt opened the pull request:
**EDIT:** [this PR is ready now](https://github.com/rust-lang/rust/pull/44042#issuecomment-333883548). I edited this post to reflect the current status of discussion, which is (apart from code review) pretty much settled.

---

This is my current progress in order to prepare stabilization of #39658. As discussed there (and in #39659), the idea is to deprecated `AsciiExt` and copy all methods to the type directly. Apparently there isn't really a reason to have those methods in an extension trait¹.

~~This is **work in progress**: copy&pasting code while slightly modifying the documentation isn't the most exciting thing to do. Therefore I wanted to already open this WIP PR after doing basically 1/4 of the job (copying methods to `&[u8]`, `char` and `&str` is still missing) to get some feedback before I continue. Some questions possibly worth discussing:~~

1. ~~Does everyone agree that deprecating `AsciiExt` is a good idea? Does everyone agree with the goal of this PR?~~ => apparently yes
2. ~~Are my changes OK so far? Did I do something wrong?~~
3. ~~The issue of the unstable-attribute is currently set to 0. I would wait until you say "Ok" to the whole thing, then create a tracking issue and then insert the correct issue id. Is that ok?~~
4. ~~I tweaked `eq_ignore_ascii_case()`: it now takes the argument `other: u8` instead of `other: &u8`. The latter was enforced by the trait. Since we're not bound to a trait anymore, we can drop the reference, ok?~~ => I reverted this, because the interface has to match the `AsciiExt` interface exactly.

¹ ~~Could it be that we can't write `impl [u8] {}`? This might be the reason for `AsciiExt`. If that is the case: is there a good reason we can't write such an impl block? What can we do instead?~~ => we couldn't at the time this PR was opened, but Simon made it possible.

/cc @SimonSapin @zackw

LukasKalbertodt commented:
I don't think the Travis-fail is my fault ... Right? :confused:

steveklabnik commented:
> /Users/travis/.travis/job_stages: line 166: shell_session_update: command not found

I do not believe so, seems... strange?

bluss commented:
“The” `impl<T> [T]` is located in liballoc (collections). I'd be thrilled if things have changed, but I think that it still restricts it so no intrinsic methods can come from core for those types. Same with `impl str`.

SimonSapin commented:
* Actual deprecations should only land in Nightly after the replacement has reached the stable channel, so that a single code base using these features can be made at any point in time to compile without warnings on all three channels.

  (Or does the compile now support something like `#[deprecated(since = "1.23")]` that is silent in earlier versions?)

* Rather than duplicating code, could one of the each pair of equivalent API call the other?

kennytm commented:
@LukasKalbertodt Any error other than `x86_64-gnu-llvm-3.7` (the first job) from the pull request CI is spurious, especially the Macs. Don't worry about that.

LukasKalbertodt commented:
@kennytm Ok, thanks!

@SimonSapin 

- Ok yes, that makes sense. I won't add the `#[deprecated]` in this PR.
- Yes, I planned to implement `AsciiExt` with the *native* methods once they are all implemented. So yes, there shouldn't really be duplicate code once I'm finished.

@bluss (and potentially everyone) I'm afraid I don't quite understand. There is this one impl-block in `liballoc/slice.rs`:

```rust
#[lang = "slice"]
#[cfg(not(test))]
impl<T> [T] { ... }
```

If I try to add `impl [u8]`, I get this error:

```
error[E0390]: only a single inherent implementation marked with `#[lang = "slice"]` is allowed for the `[T]` primitive
   --> src/liballoc/slice.rs:169:1
    |
169 | / impl [u8] {
170 | |     fn foobar(&self) {}
171 | | }
    | |_^
    |
help: consider using a trait to implement these methods
   --> src/liballoc/slice.rs:169:1
    |
169 | / impl [u8] {
170 | |     fn foobar(&self) {}
171 | | }
    | |_^
```

If I add `#[lang = "slice"]` to the new impl block, I get:

```
error[E0152]: duplicate lang item found: `slice`.
    --> src/liballoc/slice.rs:176:1
     |
176  | / impl<T> [T] {
177  | |     /// Returns the number of elements in the slice.
178  | |     ///
179  | |     /// # Example
...    |
1506 | |     }
1507 | | }
     | |_^
     |
note: first defined here.
    --> src/liballoc/slice.rs:170:1
     |
170  | / impl [u8] {
171  | |     fn foobar(&self) {}
172  | | }
     | |_^
```

So... is that just impossible? Is that what you were saying, @bluss?

SimonSapin commented:
A creative hack to work around `impl [u8] {}` not being allowed could be something like:

```rust
// The existing impl block
#[lang = "slice"]
impl<T> [T] {
    // ...

    // New methods:
    fn is_ascii(&self) where Self: AsRef<[u8]> {
        let self_as_bytes = self.as_ref();
        // ...
    }
}
```

Given the available impls of `AsRef`, that `is_ascii` method would effectively only be available for `[u8]`. However the way it shows up in rustdoc would not be pretty.

LukasKalbertodt commented:
So am I right and it is indeed not possible to add methods to `[u8]` (except for hacks, like the one @SimonSapin proposed)? If yes, does it even make sense to continue working on this PR? 

I see the following possibilities:

1. We close this PR, doing nothing ...
  a. ... and keep `AsciiExt` and eventually stabilize the `<ctype.h>` methods from #39658 
  b. ... until it is possible to add methods to `[u8]` someday in the future
2. I continue working on this PR, using Simon's hack to implement it for `[u8]` ...
  a. ... and we stabilize it exactly like that later
  b. ... and never stabilize the hack; stabilize the methods on `u8`, `char` and `str`; replace the hack once it is possible to add methods to `[u8]` directly
3. I continue working on this PR without adding the methods to `[u8]`. But as long as those methods aren't there and stabilized, we can't deprecated `AsciiExt`. 

In 1b, 2b and 3 we assume that the possibility to add methods to `[u8]` directly will be implemented eventually. **What do you think: will that be the case?** How hard is it to implement that? Are there already discussions about this topic?

SimonSapin commented:
I believe it is not possible purely in library code, just like `impl<T> [T] {}` wouldn’t be allowed if it weren’t for the special support in the compiler in the form of `#[lang = "slice"]`. Probably we could add another such special case in the compiler, but I don’t know how to do that.

aidanhs commented:
r? @sfackler since highfive flaked and this seems library related. Not sure if you'll need compiler team input though...

sfackler commented:
This might be a bit tricky to land unstable since inherent methods override trait-provided methods. So every existing use of `AsciiExt` traits is going to start hitting stability issues.

cc @rust-lang/libs.

alexcrichton commented:
The libs team discussed this during triage the other day and the conclusion was that this is good to go, @sfackler do you know what specifically needs to be updated here?

SimonSapin commented:
@alexcrichton There’s still two issues to resolve:

* Modifying a compiler to add a `bytes_slice` lang item (or something) to allow `impl [u8] {}` to exist next to `impl<T> [T] {}`. (Or mentoring from someone who’d know how to do that.)
* Decide if we’re ok with insta-stable inherent methods. If not, having them shadow trait methods is gonna make things stop compiling for users of the trait methods.

sfackler commented:
We're fine landing these as immediately stable since they're just being moved around. Not sure if we should block on figuring out the `impl [u8]` issues first or not.

SimonSapin commented:
Am I missing something? I don’t see how we can just decide to not block on having a technical solution to get past this compilation error:

```
error[E0390]: only a single inherent implementation marked with `#[lang = "slice"]` is allowed for the `[T]` primitive
```

LukasKalbertodt commented:
We *do* have the 

```rust
impl [T] {
    fn foo() where Self: AsRef<[u8]> {}
}
```

solution. But as mentioned above, it has several disadvantages and really feels like a hack. Additionally, changing it to `impl [u8]` later is probably a breaking change (I think). So I'd very much prefer to make `impl [u8]` possible before merging this change.

I could write the code to make `impl [u8]` possible, but I would need mentoring, since right now, I don't have a clue what needs to be done. Of course, I'm also fine with someone else implementing this ^_^

sfackler commented:
The not-blocking on it solution would be to keep using the extension trait for [u8].

bors commented:
:umbrella: The latest upstream changes (presumably #44678) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
ping @LukasKalbertodt would you be willing to keep up w/ this PR?

kennytm commented:
I think we are still waiting for the team's decision on whether to introduce `impl [u8]`, or workaround with `where Self: AsRef<[u8]>`, or keep the extension trait.

SimonSapin commented:
I’ll try to find a mentor in a couple weeks at the "impl days" after RustFest Zurich who could help with adding a lang item or something for `impl [u8]`.

LukasKalbertodt commented:
@alexcrichton Sure! I just wasn't sure if we decided already (what @kennytm said) ^_^

I'll rebase in the next couple of days and then move/copy the methods for `str` and `char`. Then we can see whether we want to wait for the lang item (thanks @SimonSapin! That would be awesome!) or if we want to merge without methods on `[u8]`.

arielb1 commented:
@SimonSapin 

I would be happy to help you implement `impl [u8]` in RustFest.

LukasKalbertodt commented:
I just pushed updated code. Right now, `char` and `u8` have all ascii methods directly. Also, I removed duplicate code by using the intrinsic definition in the impl of `AsciiExt`. See commit messages for more details.

Unfortunately, two problems/questions occurred:

1. Since the direct methods of `char` and `u8` shadow the `AsciiExt` methods (even if the `AsciiExt` trait is in scope), the doc-comment examples for `AsciiExt` don't work anymore. What shall I do here? I could (a) remove the examples, since `AsciiExt` will be deprecated in the future anyway, or (b) fix the examples by using explicit function calls like `AsciiExt::is_ascii(...)`.

2. Many of the ascii methods of `str` use ascii methods of `[u8]`, which I cannot implement right now, as discussed above. I also cannot use the `AsciiExt` trait in the `impl str {}` block, since `str` lives in `liballoc` and `ascii` lives in `libstd`. So I can either (a) not implement ascii methods for `str` until we can implement them for `[u8]`, or (b) not use any ascii `[u8]` methods in the `str` method definition (which would result in slightly more code, but would work).

So what do you all think about those two things?

SimonSapin commented:
Why do the doc-comment examples not work? Aren’t the new methods compatible? This whole thing was proposed on the assumption that it is not a breaking change.

As to `[u8]` I’m hopeful we can add `impl [u8] {}` and fix that next week.

LukasKalbertodt commented:
> Why do the doc-comment examples not work?

My bad, I was not precise enough. The example only fail compiling because the import `use std::ascii::AsciiExt;` is unused and this warning leads to an error. So I could simply remove all imports and it will work fine, because yes, the "new" methods are fully compatible to the old ones.

But if I just remove the import, the doc-example is not actually testing/showing the method it is attached to, but a completely different method. I thought that this is undesirable. So I guess we have three possibilities for the problem (1.):

- (a) remove the examples, since AsciiExt will be deprecated in the future anyway
- (b) fix the examples by using explicit function calls like `AsciiExt::is_ascii(...)`
- (c) fix the example by removing the imports (with the disadvantage described above)

SimonSapin commented:
I think we can move the examples to the inherent methods (or remove if they’re already there) and "soft-deprecate" `AsciiExt`: document it as deprecated in favor of the inherent methods but don’t make it warn yet.

LukasKalbertodt commented:
@SimonSapin I hope the commit I just pushed does exactly what you had in mind.

If so, I'll wait until `impl [u8] {}` is possible. Once it is, I will add the ascii methods for `str` and `[u8]`. Is this a good plan?

---

One additional question we can already discuss now: `AsciiExt` had some unstable methods (feature `ascii_ctype`, [tracking issue](https://github.com/rust-lang/rust/issues/39658)). 

1. Should this PR already stabilize those? The commits currently pushed do exactly that. If we want to wait with stabilizing (I don't know why we should, though) or if we want to create another PR to stabilize, what feature should the new methods be part of? `ascii_ctype` too or rather `ascii_ctype_methods_on_intrinsics` or something like that?

2. Since they are still unstable on `AsciiExt` should we remove them instead of deprecating (like the other methods of the trait)? I'm in favor of removing.

nagisa commented:
> SimonSapin	can someone ping the libs team for a decision on stablizing the rest of AsciiExt method (while moving them to inherent methods) ? https://github.com/rust-lang/rust/pull/44042#issuecomment-333124400

cc @rust-lang/libs

SimonSapin commented:
@LukasKalbertodt I have the lang item working at https://github.com/rust-lang/rust/compare/master...SimonSapin:impl_slice_u8

(It was easier than I expected, thanks @arielb1 !)

LukasKalbertodt commented:
@SimonSapin Sweet! Way faster than expected :)

So I should just cherry-pick your commit and include it in this PR? Or do you want to make a separate PR for this?

SimonSapin commented:
Yes. Cherry-picking, rebasing, whatever’s easier for you.

LukasKalbertodt commented:
I rebased, used the commit by @SimonSapin and implemented the remaining methods. So now this PR contains all the code it's supposed to. I will keep the "WIP" in the title though, until we decided whether or not to stabilize the ascii_ctype methods right away ([explained in this comment](https://github.com/rust-lang/rust/pull/44042#issuecomment-333124400)).

One comment about the code: since Simon added a new lang item (`slice_u8`), and since this lang item is not available in the stage0 compiler yet, I cannot add certain ascii methods in the stage0 compiler either. Due to this, there are several `cfg(stage0)` attributes all over the place. Unfortunately I also needed to add it to some imports in several files. The plan is to remove all of that stuff once stage0 is updated (how regularly does this happen?). I can already prepare a patch once this PR is merged.

And lastly: thanks already for reviewing this PR! I know it's a rather big PR although nothing particularly interesting has changed. I tried to avoid copy&paste errors and tried to adjust all the documentation for each type properly. If I can improve anything, let me know!

arielb1 commented:
@LukasKalbertodt

stage0 is updated every release, so every 6-ish weeks (there's a release in 2 weeks, so you'll probably have to wait until the next one).

SimonSapin commented:
I’ve pinged #39658 to propose a FCP for stabilization of the newer methods. In the meantime, though, it may be easier to preserve the existing stability and feature names of respective methods to allow this PR to make progress.

With that change, I think this’ll be ready for review by @rust-lang/libs.

LukasKalbertodt commented:
@SimonSapin The commit I just pushed reverts the stability status of all ascii_ctype methods to "unstable" again. I also changed the title: this PR should be ready for a PR now, as you said :)

LukasKalbertodt commented:
Now that this PR does not stabilize anything and thus should be complete, can someone remove the `S-waiting-on-author` tag and add the `S-waiting-on-review` tag? :)

sfackler commented:
This seems good to me!

@rfcbot fcp merge

dtolnay commented:
LGTM. This is only transplanting the AsciiExt methods onto the 4 types. I think we should merge this as is, then follow up in the FCP in https://github.com/rust-lang/rust/issues/39658#issuecomment-333897436 regarding changes to the methods that are unstable in this PR (which is a more complicated discussion).

@LukasKalbertodt thanks for sticking with this. This is a heroic PR.

@rfcbot reviewed

carols10cents commented:
ping @BurntSushi and @aturon for your ticky boxes [here](https://github.com/rust-lang/rust/pull/44042#issuecomment-336035680)!

carols10cents commented:
ping @BurntSushi for your ticky box [here](https://github.com/rust-lang/rust/pull/44042#issuecomment-336035680)!

alexcrichton commented:
Checking off @BurntSushi's box for him as these are unstable anyway! After which...

@bors: r+

bors commented:
:pushpin: Commit 1ff8ed5 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 1ff8ed52f4f55ba50ae0095131bd0782735f7955 -->

bors commented:
:hourglass: Testing commit 1ff8ed52f4f55ba50ae0095131bd0782735f7955 with merge 27c6ed2c1715d272a362da028d622bbb1c505dda...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5178)

bors commented:
:umbrella: The latest upstream changes (presumably #44764) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
```rust
error: unused import: `ascii`
  --> src\libstd\sys\windows\path.rs:11:5
   |
11 | use ascii::*;
   |     ^^^^^
   |
note: lint level defined here
  --> src\libstd\lib.rs:232:9
   |
232| #![deny(warnings)]
   |         ^^^^^^^^
   = note: #[deny(unused_imports)] implied by #[deny(warnings)]
error: aborting due to previous error
```

LukasKalbertodt commented:
Will rebase and fix the tests soon; hopefully in the next couple of days :)

LukasKalbertodt commented:
Rebased and fixed the mini-bug. I also searched for other places where I might have forgotten to "remove" imports, but I couldn't find any. It *should* work fine now ^_^

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit e6c0957 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton e6c09570c49ce1f715c92294034b00bddab842b4 -->

bors commented:
:hourglass: Testing commit e6c09570c49ce1f715c92294034b00bddab842b4 with merge 5787f5c86f167d9f44098314428b05bbc65462ed...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5228)

kennytm commented:
Could not compile `libstd` on Windows due to a "unused import: `ascii`" at `src\libstd\sys\windows\path.rs`.

<details>


```
error: unused import: `ascii`
  --> src\libstd\sys\windows\path.rs:12:5
   |
12 | use ascii::*;
   |     ^^^^^
   |
note: lint level defined here
  --> src\libstd\lib.rs:232:9
   |
232| #![deny(warnings)]
   |         ^^^^^^^^
   = note: #[deny(unused_imports)] implied by #[deny(warnings)]
error: aborting due to previous error
error: Could not compile `std`.
```

</details>

SimonSapin commented:
Maybe this import can be removed entirely because even `stage(0)` uses the new `std` being compiled?

LukasKalbertodt commented:
Oh, I see. Let me just try removing both windows specific imports entirely (`path` and `process`). I don't have a Windows machine setup here, so I'm kind of dependent on bors.

kennytm commented:
@bors r=alexcrichton

Let's retry!

bors commented:
:pushpin: Commit 5a8aa61 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 5a8aa61e89970e3c0b268bb3736188f45cc2500b -->

bors commented:
:hourglass: Testing commit 5a8aa61e89970e3c0b268bb3736188f45cc2500b with merge 323738f53f80995258893da455e100da3bf9632e...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5230)

kennytm commented:
```
error[E0599]: no method named `to_ascii_uppercase` found for type `sys_common::wtf8::Wtf8` in the current scope
  --> src\libstd\sys\windows\process.rs:39:35
   |
39 |         inner: s.as_inner().inner.to_ascii_uppercase()
   |                                   ^^^^^^^^^^^^^^^^^^
   |
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
           candidate #1: `use ascii::AsciiExt;`
error: aborting due to previous error
error: Could not compile `std`.
```

LukasKalbertodt commented:
Oh gosh, I should not have been clever and removed the import in `process` without really checking what's going on -.-

Will push the fix in a few minutes. I can spin up Windows tomorrow and test locally to avoid bothering bors too much. Should I do that or do you just want to r+ it again?

kennytm commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit deb7360 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton deb7360fa7838aa718b5a599460761aeb940b89e -->

bors commented:
:hourglass: Testing commit deb7360fa7838aa718b5a599460761aeb940b89e with merge 782233cedf31a8d926d5242a1da0aa0a5b1f3ecd...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/297332147?utm_source=github_status&utm_medium=notification)

LukasKalbertodt commented:
Looks like more warnings, but from Servo? I didn't know Servo was built by the built bot. I'm a bit confused, as I can't really change Servo here...? If someone could explain this, that would be nice :)

(*I'm gone for at least 12 hours tho*)

kennytm commented:
@LukasKalbertodt This is cargotest, which ensures some important packages like `xsv`, `ripgrep`, `servo` etc. are compilable with your changes so guarantee stability. The servo test was added in #44603. 

But I think the failure here is a bit unfair, as the warning is totally expected. Perhaps we can add `--cap-lints warn` to the `RUST_FLAGS` environment variable when performing the `cargo test`? (modify `src/tools/cargotest/main.rs`) 

cc @SimonSapin

SimonSapin commented:
My apologies. I’ve added Servo’s `style` crate to `cargotest` and should have anticipated this issue, since we hit exactly the same when I added it to perf.rust-lang.org.

I’ve reproduced this failure and verified that it’s fixed by this patch:

```diff
diff --git a/src/tools/cargotest/main.rs b/src/tools/cargotest/main.rs
index 4f4960efc9..a6c56a1307 100644
--- a/src/tools/cargotest/main.rs
+++ b/src/tools/cargotest/main.rs
@@ -165,6 +165,8 @@ fn run_cargo_test(cargo_path: &Path, crate_path: &Path, packages: &[&str]) -> bo
     let status = command
         // Disable rust-lang/cargo's cross-compile tests
         .env("CFG_DISABLE_CROSS_TESTS", "1")
+        // Relax #![deny(warnings)] in some crates
+        .env("RUSTFLAGS", "--cap-lints warn")
         .current_dir(crate_path)
         .status()
         .expect("");
```

I can send a PR, but it might be easier to include this patch in this PR. @LukasKalbertodt, what do you prefer?

LukasKalbertodt commented:
> This is cargotest, which ensures some important packages like `xsv`, `ripgrep`, `servo` etc. are compilable with your changes to guarantee stability. 

Interesting, I didn't know that.

> I can send a PR, but it might be easier to include this patch in this PR.

I just pushed the patch here; I guess it's easier ;-)

Thanks a lot for quickly coming up with the patch and testing it! I tested locally again, too, and no issues so far. So I guess... this time, bors might actually return with a nice result ^_^

kennytm commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit ea55596 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton ea55596d5bc29708232a0bb232bf35d5e2e6cbce -->

bors commented:
:hourglass: Testing commit ea55596d5bc29708232a0bb232bf35d5e2e6cbce with merge 94ede93467beb4ac17956845607a7e8b226dda02...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5241), [status-travis](https://travis-ci.org/rust-lang/rust/builds/297519140?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 94ede93467beb4ac17956845607a7e8b226dda02 to master...

LukasKalbertodt commented:
Yeah, finally :) My first non-tiny PR

Thanks everyone for all the kind help!

---

With this landed, I think there are the following related things:

1. Remove all the `#[cfg(stage0)]` stuff once a new stage0 is released (around the 24th, right?). I can already prepare a patch and create a PR once the stage has been updated.
2. Deprecate the `AsciiExt` trait. When do we want to do that?
3. Stabilize some of the `ascii_ctype` methods, as discussed in #39658. Do we want to do that immediately?

kennytm commented:
Step 1 will be done by a team member (e.g. #45285), so you don't need to submit that PR yourself.

SimonSapin commented:
🎉🎉🎉🎉

Thank you Lukas for pushing through all this, and sorry it took so much time and so many attempts.

The deprecation warning should wait until the replacement reaches the stable channel. That is after 1.23.0 is released.

https://github.com/rust-lang/rust/issues/39658#issuecomment-340923247 says “stabilize these methods only on the u8 and char types” so I think it’s fine to send a PR doing that for the libs team to approve.

behnam commented on src/liballoc/slice.rs:
```diff
@@ -1533,6 +1533,215 @@ impl<T> [T] {
+    /// but without allocating and copying temporaries.
+    #[stable(feature = "ascii_methods_on_intrinsics", since = "1.21.0")]
+    #[inline]
+    pub fn eq_ignore_ascii_case(&self, other: &[u8]) -> bool {
```
Any chance we can move this and similar non-alloc methods to `libcore`?

With `to_*_lowercase()/_uppercase()` methods missing in `libcore`, this is the only method available for case-insensitive string comparison, which is a common operation in many application, even embedded.

SimonSapin replied:
I agree this would be good to have, but it’s tricky. I’ve opened https://github.com/rust-lang/rust/issues/45803.

SimonSapin replied:
https://github.com/rust-lang/rust/pull/49896 does this (though https://github.com/rust-lang/rust/pull/49698 also made the Unicode `to_lowercase` and `to_uppercase` available in libcore).

albel727 commented:
And so `char::is_ascii_*` methods take `(&self)`, **unlike** `char::is_alphabetic` / `is_lowercase` / `is_whitespace` and a whole bunch of other **prior** methods, which take `(self)`, like a **Copy type** that is smaller than a reference on most arches, such as char, pretty much deserves.

Not only this is inconsistent (and possibly has worse performance too, like in the `iter()` vs `iter.cloned()` case), but it also means that `is_ascii_*` methods can't be used in shorthand way in methods like `str::split(char::is_ascii_punctuation)`, which expect `fn(char) -> bool`. 

Since `str::split()` docs even contain an example of using `str::split(char::is_uppercase)`, my sheer surprise when the same didn't work with `is_ascii_punctuation` knew no limits. 

it's of course the fault of `AsciiExt` trait, which takes `&self` in methods, because `impl AsciiExt for &'a str | &'a mut str` was harder to write than just `impl AsciiExt for str`, and so `impl AsciiExt for char` suffered from `&self` too. 

But why nobody even considered changing this before stabilizing it in 1.24? :confused:

Kimundi commented:
I think this was a unfortunate oversight. A lot of the discussion about those methods was about how to handle deprecating `AsciiExt` correctly, and moving the methods around. Since for most types keeping methods at `&self` is the only sensible way, changing to `self` might have just slipped.

I'm not sure what we can do at this point outside of possibly changing it for the next epoch, but could you open an issue for this?

LukasKalbertodt commented:
I did consider changing it to `self` (and I even did for a short time, as it seems so natural). Sadly, this would render the new functions incompatible to the `AsciiExt` ones (which, as you noted, take `&self` as parameter). I changed it back because I got compiler errors in some code which was written with `AsciiExt` in mind (I can't remember the exact code, though...). This whole change's backwards compatibility was already questionable (see #46510 for example); changing `&self` to `self` just would not be possible without breaking a lot of stuff.

kennytm commented:
@Kimundi Epoch only affects syntax and doesn't allow breaking API compatibility, meaning this change requires a real Rust 2.0 unfortunately.
